### PR TITLE
Add Axes.ecdf() method.

### PR DESCRIPTION
**Pull request by anntzer**
## PR Summary

See discussion at #16561 (attn @Wrzlprmft).  I chose to implement remove_redundant (under the name "compress") as I can see cases where it would be helpful for performance, but left "absolute" ecdfs unimplemented, as I have never seen them (and https://stats.stackexchange.com/questions/451601/what-are-absolute-ecdfs-called-if-anybody-uses-them didn't attract too much activity).

I updated the histogram_cumulative example to showcase this (as one should basically always use ecdf() instead of hist(..., cumulative=True, density=True), but let's not overdo it and immediately consider removing that, as it's probably extremely widely used).  Note that I removed the reference to astropy's bin selection examples, as it's already mentioned elsewhere (e.g. the histogram_features example) and also doesn't really apply as is to cumulative histograms.

---

Note: I am not overly convinced by silently dropping nans; we could also error on them.  (+/-inf should clearly be supported, as they have a non-ambiguous interpretation, which I have relied on before).

## PR Checklist

**Documentation and Tests**
- [ ] Has pytest style unit tests (and `pytest` passes)
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] New plotting related features are documented with examples.

**Release Notes**
- [ ] New features are marked with a `.. versionadded::` directive in the docstring and documented in `doc/users/next_whats_new/`
- [ ] API changes are marked with a `.. versionchanged::` directive in the docstring and documented in `doc/api/next_api_changes/`
- [ ] Release notes conform with instructions in  `next_whats_new/README.rst` or `next_api_changes/README.rst`

### CODE REVIEW

**Comment by oscargus** [comment]
`.. added`...?

**Comment by jklymak** [comment]
This is basically just `ax.plot(np.sort(x), np.arange(len(x)))`?   Even with weights and normalizations, this seems pretty trivial for users to compute themselves, and I'm not a huge fan of providing folks with statistical black boxes in Matplotlib core.

**Comment by anntzer** [comment]
@oscargus @story645 I have addressed your comments.

@jklymak This was already argued at https://github.com/matplotlib/matplotlib/issues/16561#issuecomment-589282916 and https://github.com/matplotlib/matplotlib/issues/16561#issuecomment-591026865: the tricky part is not the calculation (which is indeed pretty trivial), but in selecting the correct drawstyle and adding the correct point at the correct end so that the ecdf is indeed a *step* plot (see https://en.wikipedia.org/wiki/Empirical_distribution_function, https://www.itl.nist.gov/div898/software/dataplot/refman1/auxillar/ecdfplot.htm, https://fr.mathworks.com/help/stats/ecdf.html, ...) instead of connecting the points with diagonal lines (which is actually wrong).

**Comment by anntzer** [comment]
attn @Phlya @jaroslawr @mwaskom who have also commented on the original thread.

**Comment by jaroslawr** [comment]
I think this is great, histograms, box plots and violin plots are already in matplotlib and ECDFs are a broadly useful type of plot in many different application areas. I would probably use pure matplotlib if only it had this feature. Libraries like seaborn come with the cost of a whole abstraction layer over matplotlib - you often have to understand both seaborn and matplotlib and how seaborn passes arguments to matplotlib etc. Filling in a few gaps like this PR does would make using pure matplotlib much more attractive.

File: doc/api/axes_api.rst [hunk 1]
```diff
@@ -111,9 +111,10 @@ Statistics
 
    Axes.boxplot
    Axes.violinplot
+   Axes.ecdf
```
Comment (oscargus) [inline doc/api/axes_api.rst:114]:
```suggestion
   Axes.ecdf
   Axes.violinplot
```
? (As you ordered the two remaining ones...)

File: doc/users/next_whats_new/ecdf.rst [hunk 2]
```diff
@@ -0,0 +1,4 @@
+``Axes.ecdf``
+~~~~~~~~~~~~~
+A new Axes method, `~.Axes.ecdf`, allows plotting empirical cumulative
+distribution functions without any binning.
```
Comment (oscargus) [inline doc/users/next_whats_new/ecdf.rst:4]: I'd suggest adding an example plot here to make it a bit more visible. (It is not every day that we add a completely new plot type...)

Subject to getting general approval of the idea etc.

File: doc/api/axes_api.rst (hunk 1 above)
Comment (anntzer) [reply doc/api/axes_api.rst]: sure

File: doc/users/next_whats_new/ecdf.rst (hunk 2 above)
Comment (anntzer) [reply doc/users/next_whats_new/ecdf.rst:4]: sure

**Comment by Wrzlprmft** [comment]
Would it make sense to warn the users who are still using `hist(...,cumulative=True)`? At the very least, the documentation should point to the new function.

File: lib/matplotlib/axes/_axes.py [hunk 3]
```diff
@@ -7135,6 +7135,99 @@ def hist2d(self, x, y, bins=10, range=None, density=False, weights=None,
...
+
+        Other Parameters
+        ----------------
+        data : indexable object, optional
+            DATA_PARAMETER_PLACEHOLDER
+
+        **kwargs
+            Keyword arguments control the `.Line2D` properties:
+
+            %(Line2D:kwdoc)s
```
Comment (oscargus) [inline lib/matplotlib/axes/_axes.py:7183]:
```suggestion
            %(Line2D:kwdoc)%
```

Comment (anntzer) [reply lib/matplotlib/axes/_axes.py:7183]: no?

**Comment by anntzer** [comment]
@Wrzlprmft While I agree with you that there's rather few use cases of hist(..., cumulative=True) that are not better served than by ecdf(), let's first get the method in and decide on whether to include your warning later; I don't want to be derailed into a side-discussion.

File: lib/matplotlib/axes/_axes.py (hunk 3 above)
Comment (oscargus) [reply lib/matplotlib/axes/_axes.py:7183]: I may be wrong, but the current one is also wrong (or something else is wrong). Check the outcome.

Comment (anntzer) [reply lib/matplotlib/axes/_axes.py:7183]: ah, fixed

File: lib/matplotlib/axes/_axes.py [hunk 4]
```diff
@@ -7135,6 +7135,100 @@ def hist2d(self, x, y, bins=10, range=None, density=False, weights=None,
...
+                                  drawstyle="steps-pre", **kwargs)
+            line.sticky_edges.y[:] = [0, 1]
+        else:  # orientation == "horizontal":
+            if not complementary:
+                line, = self.plot([0, *cweights], [x[0], *x],
+                                  drawstyle="steps-pre", **kwargs)
+            else:
+                line, = self.plot([1, *1 - cweights], [*x, x[-1]],
+                                  drawstyle="steps-post", **kwargs)
+            print(line.get_data())
```
[the remark this answers is not in the document]
Comment (anntzer) [reply lib/matplotlib/axes/_axes.py]: woops, fixed.

**Comment by anntzer** [comment]
General agreement during call was to error on nans instead, and also error on any input that has masked values.  Also modify the docstring to suggest various ways to handle nans (ignore them, map them to +/-inf).

**Comment by jklymak** [comment]
Draft until above is implemented

**Comment by anntzer** [comment]
Done.

**Comment by jklymak** [comment]
@oscargus did you have a chance to take a second look at this?

File: lib/matplotlib/axes/_axes.py [hunk 5]
```diff
@@ -7136,6 +7136,107 @@ def hist2d(self, x, y, bins=10, range=None, density=False, weights=None,
...
         return h, xedges, yedges, pc
 
+    @_preprocess_data(replace_names=["x", "weights"], label_namer="x")
+    @_docstring.dedent_interpd
+    def ecdf(self, x, weights=None, *, complementary=False,
+             orientation="vertical", compress=False, **kwargs):
+        """
+        Compute and plot the empirical cumulative distribution function of *x*.
+
+        .. versionadded:: 3.7
```
Comment (dstansby) [inline lib/matplotlib/axes/_axes.py:7146]: This needs updating to 3.8

File: lib/matplotlib/axes/_axes.py [hunk 6]
```diff
@@ -7136,6 +7136,107 @@ def hist2d(self, x, y, bins=10, range=None, density=False, weights=None,
...
+            DATA_PARAMETER_PLACEHOLDER
+
+        **kwargs
+            Keyword arguments control the `.Line2D` properties:
+
+            %(Line2D:kwdoc)s
+
+        Returns
+        -------
+        Line2D
```
Comment (dstansby) [inline lib/matplotlib/axes/_axes.py:7189]: Link this?

File: lib/matplotlib/axes/_axes.py [hunk 7]
```diff
@@ -7136,6 +7136,107 @@ def hist2d(self, x, y, bins=10, range=None, density=False, weights=None,
...
+        if np.ma.getmask(x).any():
+            raise ValueError("ecdf() does not support masked entries")
+        x = np.asarray(x)
+        if np.isnan(x).any():
+            raise ValueError("ecdf() does not support NaNs")
+        argsort = np.argsort(x)
+        x = x[argsort]
+        if weights is None:
+            # Ensure that we end at exactly 1, avoiding floating point errors.
+            cweights = (1 + np.arange(len(x))) / len(x)
```
Comment (dstansby) [inline lib/matplotlib/axes/_axes.py:7214]: I'm a bit confused as to what `cweights` is here? perhaps a comment explaining what it is would be good?

File: lib/matplotlib/axes/_axes.py [hunk 8]
```diff
@@ -7136,6 +7136,107 @@ def hist2d(self, x, y, bins=10, range=None, density=False, weights=None,
...
+        else:
+            weights = np.take(weights, argsort)
```
Comment (dstansby) [inline lib/matplotlib/axes/_axes.py:7216]:
```suggestion
            # Sort weights
            weights = np.take(weights, argsort)
```

File: lib/matplotlib/axes/_axes.py [hunk 9]
```diff
@@ -7136,6 +7136,107 @@ def hist2d(self, x, y, bins=10, range=None, density=False, weights=None,
...
+            cweights = np.cumsum(weights / np.sum(weights))
+        if compress:
+            compress_idxs = [0, *(x[:-1] != x[1:]).nonzero()[0] + 1]
```
Comment (dstansby) [inline lib/matplotlib/axes/_axes.py:7219]:
```suggestion
            # Get indices of unique values in x
            compress_idxs = [0, *(x[:-1] != x[1:]).nonzero()[0] + 1]
```

**Review by dstansby — APPROVED** [review]
Looks good to me - I have some minor suggestions, and the version change the docs is blocking, but feel free to self-merge when that's done

**Comment by anntzer** [comment]
Thanks, I've handled all the comments.

File: doc/users/next_whats_new/ecdf.rst [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+
+.. plot::
```
Comment (tacaswell) [inline doc/users/next_whats_new/ecdf.rst:6]:
```suggestion
.. plot::
    :include-source: true
```

or we just get the image without the source!

Comment (anntzer) [reply doc/users/next_whats_new/ecdf.rst:6]: done

**Pull request merged**